# 今日记录

## 订单详情页遗留 bug

### 问题描述: 订单详情页, 加载最后一页, 一直处于 loading 状态, 明明接口已经返回

### 处理:

看代码有处理没有更多的逻辑, debugger 一下

发现根本就没有走到后面设置 setData 的逻辑, 多打几个 debugger 试一下, 发现代码在一个 if 判断的地方就没有执行下去了, 把这段代码注释就好了,
如果再多次触底, 还是会触发 loadMore, 并且也都是在这个地方不会再执行下去

代码大致是:

```js
if (
  isHave &&
  _pageList[0].cancelReasonVos &&
  _pageList[0].cancelReasonVos.length > 0
) {
  // 处理逻辑
}
// 其他逻辑
```

`_pageList` 为返回的数组, 看代码似乎是如果返回为空, 就说明没有更多了, 然后最后一页为空, 也就是 `_pageList` 为 []

按道理, `_pageList` 为空数组, `_pageList[0]` 则为 `undefined`, `_pageList[0].cancelReasonVos` 应该是报错才对, 但是并没有报错信息提示

了

测试一下:

```js
const test = [];
if (test && test[0] && test[0].a) {
  console.log("成功进入并执行了代码");
}
```

这个例子明天上班还得修改一下, 还是得找个时间在小程序 ide 上修改

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
  </head>
  <body>
    <script>
      function test() {
        const arr = [];
        if (arr && arr[0]) {
          console.log("测试一下1");
        }
        console.log("测试一下2");
      }
      test(); // 打印 测试一下2

      function test1() {
        const arr = [];
        if (arr && arr[0] && arr[0].name) {
          console.log("测试一下1");
        }
        console.log("测试一下2");
      }
      test1(); // 打印 测试一下2

      function test2() {
        const arr = [];
        if (arr[0] && arr[0].name) {
          console.log("测试一下1");
        }
        console.log("测试一下2");
      }
      test2(); // 测试一下2

      function test3() {
        const arr = [];
        if (
          true &&
          test[0].cancelReasonVos &&
          test[0].cancelReasonVos.length > 0
        ) {
          console.log("测试一下1");
        }
        console.log("测试一下2");
      }
      test3();

      // function test4() {
      //     const arr = [];
      //     if(arr && arr[0].name) {
      //         console.log('测试一下1');
      //     }
      //     console.log('测试一下2');
      // }
      // test4(); // 报错: test.html:31 Uncaught TypeError: Cannot read property 'name' of undefined
    </script>
  </body>
</html>
```

### 问题

猜测是 && 把这个错误给吃掉了, 或者说由于小程序是双线程, 报错, 但是没有通知渲染层? javascript 中的 && 的问题?

### 解决

使用`.`语法, 在有可能为 undefined 的地方加上 `?`

```js
if (
  isHave &&
  _pageList[0]?.cancelReasonVos &&
  _pageList[0]?.cancelReasonVos?.length > 0
) {
  // 处理逻辑
}
// 其他逻辑
```

## 业务梳理(门店定位)

### 1: 首页 index

在 `onLoad` 中拉取获取定位信息, 获取到定位信息后重置 `data` 中的定位数据
有一个自定义头部 `page-header` 组件, 展示门店定位，点击跳转到门店选择页面, 携带首页获取到的定位信息

### 2: page-header

监听门店信息，当门店信息发生改变了，设置新门店信息

这里使用了自定义头部 & `store-header`

自定义头部 里的内容单纯用于展示

`store-header` 的内容是搜索框等等 & 包含跳转定位门店的方法，这里点击自定义头部跳转也是使用 `selectComponent` 来选中 `store-header` 的 的方法

跳转到定位页面的时候先从 `currentPages` 中取得栈顶的 url, 拼接 `callbackUrl`, 到时候点击返回跳转回来

这里使用的是自定义头，会有一个问题，就是，低版本比如，`2.5.2` 以下的版本不支持自定义头， 以上的版本支持自定义头

头部有一个定位信息，然后搜索框，和扫码购 icon

`2.5.2` 以下的版本，以上的内容都在一行显示
`2.5.2` 以上的版本，分行显示

就会出现一个问题

头部高度不一样，那首页基本是装修组件，有些装修组件比如导航组件等等，使用的 `fix` 定位，`top` 为 0，那在支持自定义头部的版本中就会出现问题，所以在 `page-header` 中获取到自定义头部的高度，触发首页的自定义头部注册 `mount` 事件，保存一个 `pageNavigationFixedTop` 字段，用于修复使用 `fixed` 定位的装修组件的样式问题

### 3: store-header

点击跳转到选择地址页面，`options` 携带当前定位信息，拉取对应的地址信息，点击了某个地址，跳转到首页，并更新定位逻辑

## 学习点

1. `selectComponent` 添加 `class` 或者 `id`, 需要在使用的外层的组件上而不是封装的组件里

2. O2O: Online To Offline 的简称缩写, 也就是线上到线下, 线上进行打折, 促销等活动, 来吸引用户转化为自己的线下客户, 适合需要到店服务的消费

一个场景: 沃尔玛接通了视频号直播, 结果直播间里有一个商品, 秒杀了两个就没货了

产品说: 这就是 O2O 电商在直播领域的局限了。。不能保障你所在区域一定有货

3. 617 报错了，很多都报错了, 很多服务都挂掉了, 接入了视频号, 而且由于打折促销等活动, 流量进入, 商品服务挂了一片, 刚启动就挂了, 启动又挂, 然后扩容处理
